Batch inference documentation (#103)
odahu/odahu-docs: source/ref_batch.rst

###################
Batch Inference
###################

> This section describes API and protocols related to Batch inference using ODAHU

Reviewer: ```suggestion
This section describes API and protocols related to Batch inference using ODAHU.
```

ODAHU Batch Inference feature allows to user get inferences using model for large datasets that are delivered not via

> WEB API but other mechanisms.

Reviewer: ```suggestion
ODAHU Batch Inference feature allows user to get inferences using ML model for large datasets that are delivered asyncronously, not via
HTTP API, but through other mechanisms.
```

> Currently Batch Inference supports next ways to delivery data for forecasting:

Reviewer: ```suggestion
Currently Batch Inference supports the following ways to delivery data for forecasting:
```

- Object storage

  - GCS
  - S3
  - Azureblob

> For the future we consider to add ability to process data directly from Kafka topic and other async data sources.

Reviewer: ```suggestion
In future we consider to add ability to process data directly from Kafka topic and other async data sources.
```

Please also take a look at `example <https://github.com/odahu/odahu-examples/tree/feat/batch-inference-example/batch-inference>`_.

**************
API Reference
**************

=================
InferenceService
=================

> InferenceService represents next required entities:

Reviewer: ```suggestion
InferenceService represents the following required entities:
```

- Predictor docker image that contains predictor code
- Model files location on object storage (directory or .zip / .tar.gz archive)

> - Command and Arguments that describe how to execute image

Reviewer: Isn't it also responsible for input/output data location?

> When user trained a model then he should build and image with code that follows `Predictor code protocol`_ and register

Reviewer: ```suggestion
When a user trains a model then they should build an image with code that follows `Predictor code protocol`_ and register
```

this image as well as appropriate model files using ``InferenceService`` entity in ODAHU Platform.

User describes how inference should be triggered using different options in ``[].spec.triggers``.

.. openapi:: odahu-core-openapi.yaml
   :paths:
      /api/v1/batch/service

==================
InferenceJob
==================

> ``InferenceJob`` describes forecast process that was triggered by on of the triggers of ``InferenceService``.

Reviewer: ```suggestion
``InferenceJob`` describes forecast process that was triggered by one of the triggers in ``InferenceService``.
```

If ``[].spec.triggers.webhook`` is enabled then its possible to run ``InferenceJob`` by making POST request as described

> below. Be default webhook trigger is enabled. Note, that currently its the only one way to trigger jobs.

Reviewer: ```suggestion
below. By default webhook trigger is enabled. Note, that currently its the only one way to trigger jobs.
```

.. openapi:: odahu-core-openapi.yaml
   :paths:
      /api/v1/batch/job

*********************************
Predictor code protocol
*********************************

> ODAHU Platform launches docker image that provided by user as ``[].spec.image`` (InferenceService_) and guaranties the

Reviewer: ```suggestion
ODAHU Platform launches docker image provided by user as ``[].spec.image`` (InferenceService_) and guarantees the
```
